### ISSUE TITLE
Prevent full rebuilds on Kconfig changes

### ISSUE THREAD

**Issue by SebastianBoe** [zephyrproject-rtos/zephyr#4951]
With Kbuild, a Kconfig change would result in an incremental build. After the CMake migration we get a full rebuild.

Find out if the new performance is acceptable and/or if there exists any low-hanging fruit optimizations that we can take from Kbuild and re-implement in CMake.

**Comment by andrewboie** [comment]
kbuild had a clever system for enabling this. as I recall, it would create a directory structure for the kconfig space with separate files for each config option, in $OUT/include/config.

**Comment by carlescufi** [comment]
This will not work with Kconfiglib, due to the fact that it does not generate this tree in `kconfig/include` at all.

**Comment by SebastianBoe** [comment]
Until this is resolved I would recommend users that often modify Kconfig in their edit-build-debug cycle to ensure ccache is installed. Otherwise the performance will make the system unusable.

**Comment by ulfalizer** [comment]
@andrewboie 
Yep.

The way it works is that an empty file is stored in `include/config/` for each Kconfig symbol, along with a configuration file `include/config/auto.conf` that records the configuration state during the last build. When rebuilding, `.config` is compared to `auto.conf` to see which symbols changed, and the files corresponding to the changed symbols are `touch`ed to signal the changes to `make`. That avoids having to do a full rebuild whenever the configuration is changed.

There also a tool called `fixdep` that takes care of figuring out what symbols a source file depends on. It basically just greps for `CONFIG_*` stuff.

It shouldn't be that hard to add the functionality to Kconfiglib, though it requires a bit of work (would need to keep track of the `auto.conf` values separately from the `.config` values, etc.). I was thinking that you could have an API that you hand the path to an `include/config/` directory, and it would update it (or create it if it doesn't exist) and `touch` the files for the changed symbols.

**Comment by ulfalizer** [comment]
Don't know if `cmake` offers any other options by the way. You'd need some way of recording per-symbol dependencies at least. Just having them as files is natural for `make`.

**Comment by ulfalizer** [comment]
Would take less work than I first thought. Could just load `auto.conf` as a normal `.config` file, dump all the symbol values into a temporary dict, load `.config`, compare the new symbol values against the dumped values to see what changed, and flag the changes in some appropriate way. Much simpler in Python than in C. :)

Think it should be pretty trivial now, so I'm happy. Just the `cmake` end to take care of.

**Comment by ulfalizer** [comment]
Was a bit trickier than I anticipated (the devil is in the details), but it's implemented now, using the same scheme as the C tools: https://github.com/ulfalizer/Kconfiglib/commit/378dedcb38b85056991fecab24567414567724ec

**Comment by carlescufi** [comment]
@ulfalizer thanks a lot for this! I will rebase to the latest Kconfiglib after the release and we'll see how we can integrate this with CMake.

**Comment by SebastianBoe** [comment]
re-assigning to me for now, AFAICT the Kconfig part of this is done, and this is now a build-system task.

**Comment by carlescufi** [comment]
@ulfalizer is this addressed? Can we address it if not?

**Comment by ulfalizer** [comment]
@carlescufi 
Not addressed. The functionality is there in Kconfiglib, but it needs some support outside of it as well, to figure out what symbols each source file depends on.

See the docstring at https://github.com/ulfalizer/Kconfiglib/blob/673f771d2f333b5688d680eba5949f98f12c6177/kconfiglib.py#L1163 for an overview.

The tool that extracts dependencies from source files in the Linux kernel is called `fixdep`. It's just a kind of fancy `grep` for `CONFIG_*` in the source files.

**Comment by SebastianBoe** [comment]
To implement this one would need to:

Add -MD to the compile_options to have gcc spit out which header
files are used for each source file, which is trivial.

Remove CMake's autodetected dependency on autoconf.h, which I
haven't done before but I hope is possible with CMake.

Run a script on each 'gcc -MD' generated dep file to detect which
symols are used by a source file. Now, this one is tricky, as you
don't know until generation time which source files are going to
be in the build. (The source file list could be in a generator
expression).

In practice this will be difficult to implement and should
therefore be shelved until someone can figure out a simple
solution.

**Comment by carlescufi** [comment]
@SebastianBoe @ulfalizer did we ever come to a conclusion on this issue?

**Comment by ulfalizer** [comment]
@SebastianBoe 
Linux doesn't bother with `gcc -MD` when looking for symbols. It just searches the source files directly, and adds a prerequisite like `foo: include/config/my/symbol.h` if `foo.c` depends on `MY_SYMBOL`. Works fine as long as source files don't do stuff like `CONFIG_##NAME`.

See [fixdep.c](https://github.com/torvalds/linux/blob/master/scripts/basic/fixdep.c) and the docstring for `sync_deps()` in Kconfiglib for more details.

**Comment by SebastianBoe** [comment]
That is interesting. But we can't use fixdep.c because users are not required to have a host C compiler and we don't want to distribute it ourselves.

**Comment by SebastianBoe** [comment]
There is no way of solving this problem without introducing more complexity than the solution is worth.

So I am closing this as can't-fix.

If someone can figure out a reasonable way to solve this then feel free to comment.

**Comment by AndreyDodonov-EH** [comment]
Even with `ccache` changing prj.conf is super-slow - ccache falls back to preprocess mode for every TU, since they all reference autoconf.h

So I made a script which without touching Zephyr speeds up the build: 
[fast-kconfig-rebuild](https://github.com/AndreyDodonov-EH/fast-kconfig-rebuild)

It's dirty - it checks for changes in configs, then if only int/hex/strings were changed, it patches autoconf.h/.config/configc.c,
and restores mtimes - this way Ninja doesn't notice the change. It also greps for object files which are influenced by the config, so they get deleted and do get rebuilt. So dirty AF, but it speeds up my development builds x10 (40s to 4s on ca. 560 translation units)
Also wrapped as `west` command

Feedback is welcome - parts of it are AI (specifically python part inside), so treat with care.
Still, I hope it helps somebody to save time.
